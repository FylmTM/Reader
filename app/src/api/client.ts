type Response<R> =
  | { status: "ok"; payload: R }
  | { status: "error"; message: string };

function post<T, R>(path: string, data: T): Promise<R> {
  return fetch(path, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  })
    .then(response => {
      return response.json() as Promise<Response<R>>;
    })
    .then(result => {
      switch (result.status) {
        case "error":
          throw Error(result.message);
        case "ok":
          return result.payload;
      }
    });
}

function get<R>(path: string): Promise<R> {
  return fetch(path, { method: "GET" })
    .then(response => {
      return response.json() as Promise<Response<R>>;
    })
    .then(result => {
      switch (result.status) {
        case "error":
          throw Error(result.message);
        case "ok":
          return result.payload;
      }
    });
}

export default {
  get,
  post,
};
