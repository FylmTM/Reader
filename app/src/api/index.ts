import { CategoriesWithFeeds, Post, User } from "../domain";

const DELAY = 200;

function getCurrentUser(): Promise<User> {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve({
        id: 1,
        username: "me"
      });
    }, DELAY);
  });
}

function login(): Promise<User> {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve({
        id: 1,
        username: "me"
      });
    }, DELAY);
  });
}

function logout(): Promise<void> {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, DELAY);
  });
}

function getCategoriesWithFeeds(): Promise<CategoriesWithFeeds> {
  return new Promise<CategoriesWithFeeds>((resolve, reject) => {
    setTimeout(() => {
      resolve([
        {
          category: {
            id: 1,
            name: "General"
          },
          feeds: [
            {
              id: 1,
              kind: "RSS",
              title: "Hacker News",
              link: "https://news.ycombinator.com/",
              feed: "https://news.ycombinator.com/rss"
            },
            {
              id: 2,
              kind: "RSS",
              title: "vc.ru",
              link: "https://vc.ru/",
              feed: "https://vc.ru/rss/all"
            }
          ]
        },
        {
          category: {
            id: 2,
            name: "Gaming - Best Gaming Category in the whole World"
          },
          feeds: [
            {
              id: 3,
              kind: "RSS",
              title: "StopGame - Best Gaming Website in the World",
              link: "https://stopgame.ru",
              feed: "https://rss.stopgame.ru/rss_all.xml"
            }
          ]
        }
      ]);
    }, DELAY);
  });
}

function randomInt(min: number, max: number) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function getPosts(): Promise<Array<Post>> {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve(
        Array(15)
          .fill(".")
          .map((_, i) => {
            return {
              id: i + 1,
              is_read: false,
              is_read_later: false,
              feed: {
                id: 1,
                category_id: 1,
                title: "Hacker News"
              },
              link: "http://example.com",
              title:
                randomInt(0, 1) === 1
                  ? "World is finally living on mars."
                  : "World is falling down to the works of ancient civilization who is going to do a revolution.",
              summary:
                randomInt(0, 1) === 1
                  ? Array(randomInt(1, 100))
                      .fill("lorem")
                      .join(" ")
                  : undefined,
              date: `${randomInt(1, 10)}h`,
              media:
                randomInt(0, 1) === 1
                  ? {
                      type: "image/jpeg",
                      // link: "http://localhost:3000/android-chrome-512x512.png"
                      link:
                        "https://stopga.me/images/video/2019/12/07/M8MFv_nN.jpg"
                    }
                  : undefined,
              comments_link:
                randomInt(0, 1) === 1
                  ? "http://example.com/#comments"
                  : undefined
            };
          })
      );
    }, DELAY);
  });
}

function markAsRead(postId: number, isRead: Boolean): Promise<void> {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, DELAY);
  });
}

function markAsReadLater(postId: number, isRead: Boolean): Promise<void> {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, DELAY);
  });
}

export default {
  getCurrentUser,
  login,
  logout,
  getCategoriesWithFeeds,
  getPosts,
  markAsRead,
  markAsReadLater
};
