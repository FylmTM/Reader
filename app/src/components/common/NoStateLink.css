span.link:hover {
    text-decoration: underline;
}
