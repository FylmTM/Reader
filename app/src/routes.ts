export const ALL_ROUTE = "/";

export const CATEGORY_ROUTE = "/category/:categoryId";
export const CATEGORY_POST_ROUTE = "/category/:categoryId/post/:postId";
export const FEED_ROUTE = "/feed/:feedId";
export const FEED_POST_ROUTE = "/feed/:feedId/post/:postId";
