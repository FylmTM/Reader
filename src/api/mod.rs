use crate::error::Error;
use rocket_contrib::json::Json;

pub mod authentication;
pub mod catchers;
pub mod endpoints;

#[derive(Serialize, Deserialize)]
pub struct ApiResponse<T> {
    status: String,
    payload: T,
}

impl<T> ApiResponse<T> {
    fn with(payload: T) -> ApiResponse<T> {
        ApiResponse {
            status: "ok".to_string(),
            payload,
        }
    }
}

type Response<T> = Result<Json<ApiResponse<T>>, Error>;

fn ok<T>(data: T) -> Response<T> {
    Ok(Json(ApiResponse::with(data)))
}
