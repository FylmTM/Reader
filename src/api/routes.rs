use std::ops::Add;

use rocket::http::{Cookie, Cookies};
use rocket_contrib::json::Json;

use crate::api::{ok, Response};
use crate::db::{self, Queries};

#[derive(Serialize, Deserialize)]
pub struct ApiKey {
    pub api_key: String,
}

#[post("/api/v1/authenticate", data = "<api_key>")]
pub fn authenticate(
    api_key: Json<ApiKey>,
    conn: db::PoolConnection,
    mut cookies: Cookies,
) -> Response<db::User> {
    let api_key = api_key.0.api_key;
    let user = conn.find_user_by_api_key(&api_key)?;
    cookies.add(
        Cookie::build("api_key", api_key)
            .path("/")
            .http_only(true)
            .expires(time::now().add(time::Duration::days(365)))
            .finish(),
    );

    ok(user)
}

#[get("/api/v1/current_user")]
pub fn current_user(user: db::User) -> Response<db::User> {
    ok(user)
}
