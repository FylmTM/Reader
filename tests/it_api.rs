#![feature(proc_macro_hygiene, decl_macro)]
#[macro_use]
extern crate rocket;

#[macro_use]
extern crate insta;

use common::client::{self, RequestOperations, ResponseOperations};
use reader::api::routes::*;

pub mod common;

#[test]
fn test_auth_login_failed() {
    let client = client::get();
    let mut response = client
        .post("/api/v1/auth/login")
        .json(ApiKey {
            api_key: "invalid".to_string(),
        })
        .dispatch();

    assert_json_snapshot!(response.json_entity())
}

#[test]
fn test_auth_login_success() {
    let client = client::get();
    let mut response = client
        .post("/api/v1/auth/login")
        .json(ApiKey {
            api_key: "api_key".to_string(),
        })
        .dispatch();

    assert_json_snapshot!(response.json_entity());
    assert_eq!(response.cookies().len(), 1);
}

#[test]
fn test_auth_logout_success() {
    let client = client::authenticated();

    let mut response = client.post("/api/v1/auth/logout").dispatch();
    assert_eq!(response.entity(), client::api_ok(()));
    assert_eq!(response.cookies().len(), 1);

    let mut response = client.get("/api/v1/current_user").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_current_user_failed_if_not_authenticated() {
    let client = client::get();

    let mut response = client.get("/api/v1/current_user").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_current_user() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/current_user").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_categories_with_feeds() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/categories_with_feeds").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_posts_all() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/posts").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_posts_read_later() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/posts?is_read_later=true").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_posts_category() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/posts?category_id=1").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_posts_feed() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/posts?feed_id=1").dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_post_content() {
    let client = client::authenticated();

    let mut response = client.get("/api/v1/posts/1/content").dispatch();
    assert_json_snapshot!(response.json_entity());
}
