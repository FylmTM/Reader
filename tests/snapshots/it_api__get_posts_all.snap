---
source: tests/it_api.rs
expression: response.json_entity()
---
[
  "200 OK",
  {
    "payload": {
      "has_next_page": true,
      "items": [
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-12-03T00:00:00Z",
          "feed_id": 9,
          "feed_title": "Rust Blog",
          "id": 164,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.rust-lang.org/2019/12/03/survey-launch.html",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Launching the 2019 State of Rust Survey"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-11-07T00:00:00Z",
          "feed_id": 9,
          "feed_title": "Rust Blog",
          "id": 163,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.rust-lang.org/2019/11/07/Rust-1.39.0.html",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Announcing Rust 1.39.0"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-11-07T00:00:00Z",
          "feed_id": 9,
          "feed_title": "Rust Blog",
          "id": 162,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.rust-lang.org/2019/11/07/Async-await-stable.html",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Async-await on stable Rust!"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-11-01T00:00:00Z",
          "feed_id": 9,
          "feed_title": "Rust Blog",
          "id": 161,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.rust-lang.org/2019/11/01/nll-hard-errors.html",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Completing the transition to the new borrow checker"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-10-29T00:00:00Z",
          "feed_id": 9,
          "feed_title": "Rust Blog",
          "id": 160,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.rust-lang.org/2019/10/29/A-call-for-blogs-2020.html",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "A call for blogs 2020"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-10-15T00:00:00Z",
          "feed_id": 9,
          "feed_title": "Rust Blog",
          "id": 159,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.rust-lang.org/2019/10/15/Rustup-1.20.0.html",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Announcing Rustup 1.20.0"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-10-03T00:00:00Z",
          "feed_id": 9,
          "feed_title": "Rust Blog",
          "id": 158,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.rust-lang.org/2019/10/03/inside-rust-blog.html",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Announcing the Inside Rust blog"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-09-30T00:00:00Z",
          "feed_id": 9,
          "feed_title": "Rust Blog",
          "id": 157,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.rust-lang.org/2019/09/30/Security-advisory-for-cargo.html",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Security advisory for Cargo"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-09-30T00:00:00Z",
          "feed_id": 9,
          "feed_title": "Rust Blog",
          "id": 156,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.rust-lang.org/2019/09/30/Async-await-hits-beta.html",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Async-await hits beta!"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-09-26T00:00:00Z",
          "feed_id": 9,
          "feed_title": "Rust Blog",
          "id": 155,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.rust-lang.org/2019/09/26/Rust-1.38.0.html",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Announcing Rust 1.38.0"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-12-05T10:41:54Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 154,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/twitter-memes-challenge/",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Challenge: find Twitter memes with suffix arrays"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-12-03T14:57:10Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 153,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/2019/12/03/solutions-to-the-tiny-window-manager-challenge/",
          "media_link": "https://jvns.ca/images/bouncewm.gif",
          "media_type": "image/*",
          "read_date": null,
          "summary": null,
          "title": "Solutions to the tiny window manager challenge"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-11-25T18:03:40Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 152,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/2019/11/25/challenge--make-a-bouncy-window-manager/",
          "media_link": "https://jvns.ca/images/bouncewm.gif",
          "media_type": "image/*",
          "read_date": null,
          "summary": null,
          "title": "Challenge: Write a bouncy window manager"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-11-20T20:33:20Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 151,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/2019/11/20/what-makes-a-programming-exercise-good/",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "What makes a programming exercise good?"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-11-18T13:34:29Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 150,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/2019/11/18/how-containers-work--overlayfs/",
          "media_link": "https://jvns.ca/images/overlay.jpeg",
          "media_type": "image/*",
          "read_date": null,
          "summary": null,
          "title": "How containers work: overlayfs"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-11-18T10:47:41Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 149,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/2019/11/18/some-notes-on-vector-drawing-apps/",
          "media_link": "https://jvns.ca/images/notability.jpg",
          "media_type": "image/*",
          "read_date": null,
          "summary": null,
          "title": "Some notes on vector drawing apps"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-10-29T10:51:14Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 148,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/2019/10/28/some-research-on-shipping-print-zines/",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Some research on shipping print zines"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-10-28T11:19:19Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 147,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/2019/10/28/sqlite-is-really-easy-to-compile/",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "SQLite is really easy to compile"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-10-21T16:15:19Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 146,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/2019/10/21/print-collection-of-my-first-7-zines/",
          "media_link": "https://jvns.ca/images/toolbox-inbox.jpeg",
          "media_type": "image/*",
          "read_date": null,
          "summary": null,
          "title": "Your Linux Toolbox: a box set of my free zines"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-10-03T10:24:53Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 145,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/2019/10/03/sql-queries-don-t-start-with-select/",
          "media_link": "https://jvns.ca/images/sql-queries.jpeg",
          "media_type": "image/*",
          "read_date": null,
          "summary": null,
          "title": "SQL queries don't start with SELECT"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-10-01T12:53:58Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 144,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/2019/10/01/zine-revenue-2019/",
          "media_link": "https://jvns.ca/images/revenue-by-month.png",
          "media_type": "image/*",
          "read_date": null,
          "summary": null,
          "title": "Zine revenue for 2019"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-09-30T12:35:51Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 143,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/2019/09/30/notes-on-building-sql-exercises/",
          "media_link": "https://jvns.ca/images/aggregation-challenge.png",
          "media_type": "image/*",
          "read_date": null,
          "summary": null,
          "title": "Notes on building SQL exercises"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-09-13T11:05:15Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 142,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/2019/09/13/a-year-explaining-computer-things/",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Taking a year to explain computer things"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-09-12T12:13:27Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 141,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/2019/09/12/new-zine-on-http/",
          "media_link": "https://jvns.ca/images/http-zine-cover.png",
          "media_type": "image/*",
          "read_date": null,
          "summary": null,
          "title": "New zine: HTTP: Learn your browser's language!"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-09-06T15:56:37Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 140,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/2019/09/06/how-to-put-an-html-page-on-the-internet/",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "How to put an HTML page on the internet"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-09-01T09:02:43Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 139,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/2019/09/01/ways-to-write-zines-without-fancy-tools/",
          "media_link": "https://jvns.ca/images/drawing-status-codes.png",
          "media_type": "image/*",
          "read_date": null,
          "summary": null,
          "title": "How to write zines with simple tools"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-08-30T09:25:15Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 138,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/2019/08/30/git-exercises--navigate-a-repository/",
          "media_link": "https://wizardzines.com/zines/oh-shit-git/samples/ohshit-commit.png",
          "media_type": "image/*",
          "read_date": null,
          "summary": null,
          "title": "git exercises: navigate a repository"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-08-27T09:29:38Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 137,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/2019/08/27/curl-exercises/",
          "media_link": "https://jvns.ca/images/curl.jpeg",
          "media_type": "image/*",
          "read_date": null,
          "summary": null,
          "title": "curl exercises"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-06-28T18:46:02Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 136,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/brag-documents/",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Get your work recognized: write a brag document"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-06-23T18:48:35Z",
          "feed_id": 8,
          "feed_title": "Julia Evans",
          "id": 135,
          "is_read": false,
          "is_read_later": false,
          "link": "https://jvns.ca/blog/2019/06/23/a-few-debugging-resources/",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "What does debugging a program look like?"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-11-20T11:00:00Z",
          "feed_id": 7,
          "feed_title": "Golang Blog",
          "id": 134,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.golang.org/survey2019",
          "media_link": null,
          "read_date": null,
          "summary": "Since 2016, thousands of Gophers around the world have helped the Go project by sharing your thoughts via our annual Go Developer Survey. Your feedback has played an enormous role in driving changes to our language, ecosystem, and community, including the gopls language server, new error-handling mechanics, the module mirror, and so much more from the latest Go 1.13 release. And of course, we publicly share each year's results, so we can all benefit from the community's insights.\n",
          "title": "Announcing the 2019 Go Developer Survey"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-11-13T11:00:00Z",
          "feed_id": 7,
          "feed_title": "Golang Blog",
          "id": 133,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.golang.org/go.dev",
          "media_link": "https://blog.golang.org/go.dev/home.png",
          "media_type": "image/*",
          "read_date": null,
          "summary": "Over the last two years,\nas we’ve spoken with users at companies of all sizes,\nwe’ve heard three questions repeatedly:\nwho else is using Go,\nwhat do they use it for,\nand how can I find useful Go packages?\n",
          "title": "Go.dev: a new hub for Go developers"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-11-08T11:00:00Z",
          "feed_id": 7,
          "feed_title": "Golang Blog",
          "id": 132,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.golang.org/10years",
          "media_link": "https://blog.golang.org/10years/gopher10th-small.jpg",
          "media_type": "image/*",
          "read_date": null,
          "summary": "Happy birthday, Go!\n",
          "title": "Go Turns 10"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-11-07T11:00:00Z",
          "feed_id": 7,
          "feed_title": "Golang Blog",
          "id": 131,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.golang.org/v2-go-modules",
          "media_link": null,
          "read_date": null,
          "summary": "This post is part 4 in a series.\n",
          "title": "Go Modules: v2 and Beyond"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-10-17T11:00:00Z",
          "feed_id": 7,
          "feed_title": "Golang Blog",
          "id": 130,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.golang.org/go1.13-errors",
          "media_link": null,
          "read_date": null,
          "summary": "Go’s treatment of errors as values\nhas served us well over the last decade. Although the standard library’s support\nfor errors has been minimal—just the errors.New and fmt.Errorf functions,\nwhich produce errors that contain only a message—the built-in error interface\nallows Go programmers to add whatever information they desire. All it requires\nis a type that implements an Error method:\n",
          "title": "Working with Errors in Go 1.13"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-09-26T11:00:00Z",
          "feed_id": 7,
          "feed_title": "Golang Blog",
          "id": 129,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.golang.org/publishing-go-modules",
          "media_link": null,
          "read_date": null,
          "summary": "This post is part 3 in a series.\n",
          "title": "Publishing Go Modules"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-09-03T11:00:00Z",
          "feed_id": 7,
          "feed_title": "Golang Blog",
          "id": 128,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.golang.org/go1.13",
          "media_link": null,
          "read_date": null,
          "summary": "Today the Go team is very happy to announce the release of Go 1.13. You can get it from the download page.\n",
          "title": "Go 1.13 is released"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-08-29T11:00:00Z",
          "feed_id": 7,
          "feed_title": "Golang Blog",
          "id": 127,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.golang.org/module-mirror-launch",
          "media_link": "https://blog.golang.org/module-mirror-launch/proxy-protocol.png",
          "media_type": "image/*",
          "read_date": null,
          "summary": "We are excited to share that our module mirror,\nindex, and\nchecksum database are now production ready! The go command\nwill use the module mirror and checksum database by default for\nGo 1.13 module users.  See\nproxy.golang.org/privacy for privacy\ninformation about these services and the\ngo command documentation\nfor configuration details, including how to disable the use of these servers or\nuse different ones.  If you depend on non-public modules, see the\ndocumentation for configuring your environment.\n",
          "title": "Module Mirror and Checksum Database Launched"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-08-21T11:00:00Z",
          "feed_id": 7,
          "feed_title": "Golang Blog",
          "id": 126,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.golang.org/migrating-to-go-modules",
          "media_link": null,
          "read_date": null,
          "summary": "This post is part 2 in a series.\n",
          "title": "Migrating to Go Modules"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-08-15T11:00:00Z",
          "feed_id": 7,
          "feed_title": "Golang Blog",
          "id": 125,
          "is_read": false,
          "is_read_later": false,
          "link": "https://blog.golang.org/contributors-summit-2019",
          "media_link": "https://blog.golang.org/contributors-summit-2019/group.jpg",
          "media_type": "image/*",
          "read_date": null,
          "summary": "For the third year in a row, the Go team and contributors convened\nthe day before GopherCon to discuss and plan for the future of the Go project.\nThe event included self-organizing into breakout groups,\na town-hall style discussion about the proposal process in the morning,\nand afternoon break-out roundtable discussions\nbased on topics our contributors chose.\nWe asked five contributors to write about their experience\nin various discussions at this year’s summit.\n",
          "title": "Contributors Summit 2019"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-12-11T03:17:37Z",
          "feed_id": 6,
          "feed_title": "GitHub Rocket Releases",
          "id": 124,
          "is_read": false,
          "is_read_later": false,
          "link": "https://github.com/SergioBenitez/Rocket/releases/tag/async-pre-rebase3",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "async-pre-rebase3"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-06-29T01:26:16Z",
          "feed_id": 6,
          "feed_title": "GitHub Rocket Releases",
          "id": 123,
          "is_read": false,
          "is_read_later": false,
          "link": "https://github.com/SergioBenitez/Rocket/releases/tag/v0.4.2",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Rocket v0.4.2"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2019-05-11T23:56:52Z",
          "feed_id": 6,
          "feed_title": "GitHub Rocket Releases",
          "id": 122,
          "is_read": false,
          "is_read_later": false,
          "link": "https://github.com/SergioBenitez/Rocket/releases/tag/v0.4.1",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Rocket v0.4.1"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2018-12-08T16:50:40Z",
          "feed_id": 6,
          "feed_title": "GitHub Rocket Releases",
          "id": 121,
          "is_read": false,
          "is_read_later": false,
          "link": "https://github.com/SergioBenitez/Rocket/releases/tag/v0.4.0",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Rocket v0.4.0"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2018-12-01T05:25:08Z",
          "feed_id": 6,
          "feed_title": "GitHub Rocket Releases",
          "id": 120,
          "is_read": false,
          "is_read_later": false,
          "link": "https://github.com/SergioBenitez/Rocket/releases/tag/v0.4.0-rc.2",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Rocket v0.4-rc.2"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2018-10-31T22:09:18Z",
          "feed_id": 6,
          "feed_title": "GitHub Rocket Releases",
          "id": 119,
          "is_read": false,
          "is_read_later": false,
          "link": "https://github.com/SergioBenitez/Rocket/releases/tag/v0.4.0-rc.1",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Rocket v0.4-rc.1"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2018-09-30T01:05:28Z",
          "feed_id": 6,
          "feed_title": "GitHub Rocket Releases",
          "id": 118,
          "is_read": false,
          "is_read_later": false,
          "link": "https://github.com/SergioBenitez/Rocket/releases/tag/v0.3.17",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Rocket v0.3.17"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2018-08-24T07:33:16Z",
          "feed_id": 6,
          "feed_title": "GitHub Rocket Releases",
          "id": 117,
          "is_read": false,
          "is_read_later": false,
          "link": "https://github.com/SergioBenitez/Rocket/releases/tag/v0.3.16",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Rocket v0.3.16"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2018-07-16T08:23:30Z",
          "feed_id": 6,
          "feed_title": "GitHub Rocket Releases",
          "id": 116,
          "is_read": false,
          "is_read_later": false,
          "link": "https://github.com/SergioBenitez/Rocket/releases/tag/v0.3.15",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Rocket v0.3.15"
        },
        {
          "category_id": 2,
          "comments_link": null,
          "date": "2018-06-23T06:39:06Z",
          "feed_id": 6,
          "feed_title": "GitHub Rocket Releases",
          "id": 115,
          "is_read": false,
          "is_read_later": false,
          "link": "https://github.com/SergioBenitez/Rocket/releases/tag/v0.3.14",
          "media_link": null,
          "read_date": null,
          "summary": null,
          "title": "Rocket v0.3.14"
        }
      ]
    },
    "status": "ok"
  }
]
